src/interface/state_descriptor.cpp

```diff
@@ -0,0 +1,347 @@
+//========================================================================================
+// (C) (or copyright) 2020. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#include <iomanip>
+#include <iostream>
+#include <sstream>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "basic_types.hpp"
+#include "interface/metadata.hpp"
+#include "interface/state_descriptor.hpp"
+
+namespace parthenon {
+
+// Helper class for ResolvePackages
+struct DependencyTracker {
+  std::unordered_set<std::string> provided_vars;
+  std::unordered_set<std::string> depends_vars;
+
+  Dictionary<int> overridable_vars;
+  Dictionary<std::unordered_map<int, bool>> sparse_added_provides;
+  Dictionary<std::unordered_map<int, bool>> sparse_added_depends;
+  Dictionary<std::vector<Metadata>> overridable_meta;
+
+  template <typename AdderPrivate, typename AdderProvides>
+  void Sort(const std::string &package, const std::string &var, const Metadata &metadata,
```

Comment:
Is this really sorting?

Reply:
It's sorting into categories. I guess a better name would be `Categorize`.